solana/programs/example-native-token-transfers/src/instructions/release_inbound.rs

```diff
@@ -78,38 +89,65 @@ pub fn release_inbound_mint(
     }
 
     assert!(inbox_item.release_status == ReleaseStatus::Released);
-    match ctx.accounts.common.config.mode {
-        Mode::Burning => token_interface::mint_to(
-            CpiContext::new_with_signer(
-                ctx.accounts.common.token_program.to_account_info(),
-                token_interface::MintTo {
-                    mint: ctx.accounts.common.mint.to_account_info(),
-                    to: ctx.accounts.common.recipient.to_account_info(),
-                    authority: ctx.accounts.common.token_authority.clone(),
-                },
-                &[&[
-                    crate::TOKEN_AUTHORITY_SEED,
-                    &[ctx.bumps.common.token_authority],
-                ]],
-            ),
-            inbox_item.amount,
+
+    // NOTE: minting tokens is a two-step process:
+    // 1. Mint tokens to the custody account
+    // 2. Transfer the tokens from the custody account to the recipient
+    //
+    // This is done to ensure that if the token has a transfer hook defined, it
+    // will be called after the tokens are minted.
+    // Unfortunately the Token2022 program doesn't trigger transfer hooks when
+    // minting tokens, so we have to do it "manually" via a transfer.
+    //
+    // If we didn't do this, transfer hooks could be bypassed by transferring
+    // the tokens out through NTT first, then back in to the intended recipient.
+    //
+    // The [`transfer_burn`] function operates in a similar way
+    // (transfer to custody from sender, *then* burn).
+
+    // Step 1: mint tokens to the custody account
+    token_interface::mint_to(
+        CpiContext::new_with_signer(
+            ctx.accounts.common.token_program.to_account_info(),
+            token_interface::MintTo {
+                mint: ctx.accounts.common.mint.to_account_info(),
+                to: ctx.accounts.common.custody.to_account_info(),
+                authority: ctx.accounts.common.token_authority.clone(),
+            },
+            &[&[
+                crate::TOKEN_AUTHORITY_SEED,
+                &[ctx.bumps.common.token_authority],
+            ]],
         ),
-        Mode::Locking => Err(NTTError::InvalidMode.into()),
```

Comment:
Note to self/other reviews: 
Deleting this is OK because it's guaranteed by the new constraint on the composite account `ReleaseInboundMint`